tests/5.0/target/test_target_collapse.c

```diff
@@ -0,0 +1,62 @@
+//===------ test_target_collapse.c --------------------------------------===//
+//
+// OpenMP API Version 5.0 Nov 2018
+//
+// Tests the target directive with collapse clause.
+//
+//===------------------------------------------------------------------------===//
+
+
+#include <stdio.h>
+#include "omp.h"
+#include "ompvv.h"
+
+#define N 5
+int Runtst(int gpu) {
+  int A[N], B[N] = {0};
+  int errors = 0;
+  for (int i = 0; i < N; ++i) {
+    A[i] = 0;
+    B[i] = i + 1;
+  }
+  int ThrdTrack[N * N + N] = {0}; // an array of 30 elements
+#pragma omp target data map(tofrom: A, B, ThrdTrack) device(gpu)
+#pragma omp target parallel for collapse(2) shared(A, B, ThrdTrack) device(gpu)
+  for (int i = 0; i < N; ++i) {
+    for (int j = 0; j < N; ++j) {
+      int ThrdId = omp_get_thread_num();
+      ThrdTrack[ThrdId] = ThrdId;
+#pragma omp atomic
+      A[i] += B[j];
+    }
+  }
+
+  // verifying the result
+  for (int i = 0; i < N; ++i) {
+    if (A[i] != N * (N + 1)/2) {
+      errors++;
+    }
+    if (ThrdTrack[i] != i) {
+      errors++;
+    }
+  }
+  // since only 25 threads are expected to get launched the remaining 5 places
+  // of ThrdTrack[] array should still have zeros
+  for (int i = (N*N); i < (N*N+N); ++i) {
```

Comment:
This may work correctly only if `num_threads(N*N)` clause is explicitly added to the parallel construct, and even with `num_threads(N*N)`, the behavior is not guaranteed since `num_threads(N*N)` does not guarantee that the actual number of threads is N*N (`num_threads` clause defines the upper limit on the number of threads in a parallel region).